### PR TITLE
Lite: Websocket queueing

### PR DESCRIPTION
**Pull request by whitphx**
## Description

Please include a concise summary, in clear English, of the changes in this pull request. If it closes an issue, please mention it here.

Closes: #(issue)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by whitphx** [comment]
Note:
`Queue.access_token` should be removed.

**Comment by whitphx** [comment]
The queue works on Gradio-lite.
CleanShot 2023-09-11 at 18 28 04

**Comment by aliabd** [comment]
@whitphx like we discussed in sync, I ran into a few issues: 

- Demos with queue enabled are hanging forever, but there's no error on the console. weirdly, only demos with `demo.load` are working for me. Everything else hangs.

Here's two simple examples:

```python
import gradio as gr

demo = gr.Interface(lambda x:x, "text", "text")

if __name__ == "__main__":
    demo.queue().launch()
```

```python
import gradio as gr

def greet(inp):
	return "hello " + inp + "!"

with gr.Blocks() as demo:
	text = gr.Textbox()
	output = gr.Textbox()
	button = gr.Button("Submit")
	button.click(greet, inputs=text, outputs=output)


if __name__ == "__main__":
	demo.queue().launch()

```

- There's an issue with executing files on the App.svelte. If I make a change on the editor and click execute, I get this error: 

```
Uncaught (in promise) PythonError: Traceback (most recent call last):
  File "<exec>", line 120, in _run_script
  File "app.py", line 12, in <module>
    dep = demo.load(greet_with_time, None, text, every=1)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/gradio/blocks.py", line 1704, in load
    dep, dep_index = self.set_event_trigger(
                     ^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/gradio/blocks.py", line 326, in set_event_trigger
    "outputs": [block._id for block in outputs],
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/gradio/blocks.py", line 326, in <listcomp>
    "outputs": [block._id for block in outputs],
                ^^^^^^^^^
AttributeError: 'Markdown' object has no attribute '_id'

    at new_error (pyodide.asm.js:9:12519)
    at pyodide.asm.wasm
    at pyodide.asm.wasm
    at Module.callPyObjectKwargs (pyodide.asm.js:9:81856)
    at Module.callPyObject (pyodide.asm.js:9:82066)
    at Function.apply (pyodide.asm.js:9:97147)
    at Object.apply (pyodide.asm.js:9:95381)
    at self.onmessage (webworker.js?type=classic&worker_file:333:795)

```

**Comment by whitphx** [comment]
@aliabd Thank you! Will work on these!

**Comment by whitphx** [comment]
Note:

### About the first problem
> Demos with queue enabled are hanging forever, but there's no error on the console. weirdly, only demos with demo.load are working for me. Everything else hangs.

This is because `startup-events` is not called now in the Wasm env:
https://github.com/gradio-app/gradio/blob/main/gradio/blocks.py#L2041-L2046

-> I will fix here to call `startup_events()`.

### About the second problem
> There's an issue with executing files on the App.svelte. If I make a change on the editor and click execute, I get this error: AttributeError: 'Markdown' object has no attribute '_id'

`id` is set here: https://github.com/gradio-app/gradio/blob/main/gradio/blocks.py#L141
Looks like multi-thread, async, or some Wasm-specific nature affect this with a naive usage of a global `Context` class. Luckily, found https://github.com/gradio-app/gradio/pull/5625 solves this exact problem 🎉

-> I will merge the latest `main` which already has this fix.

File: gradio/blocks.py [hunk 1]
```diff
@@ -2292,6 +2299,20 @@ def close(self, verbose: bool = True) -> None:
...
+                # so they will be cancelled by `self.server.close()` above.
+                # However, in the Wasm env, we don't have the `server` and
+                # all async tasks are running in the same event loop, `pyodide.webloop.WebLoop` in the main thread,
+                # so we have to cancel them manually.
+                for task in asyncio.all_tasks():
+                    task_name = task.get_name()
+                    if re.match(
+                        r"\w+_\d+", task_name
+                    ):  # This regex matches the task names set by `utils.set_task_name()`
+                        task.cancel()
```
Comment (whitphx) [inline gradio/blocks.py:2315]: This code is indented to cancel the tasks created at the following places, which normally run in an event loop in the server thread and stopped by `self.server.cancel()`, however, in the case of Wasm, run in an event loop in the main thread.
* https://github.com/gradio-app/gradio/blob/ea0e00b207b4b90a10e9d054c4202d4e705a29ba/gradio/routes.py#L573-L576
* https://github.com/gradio-app/gradio/blob/ea0e00b207b4b90a10e9d054c4202d4e705a29ba/gradio/queueing.py#L136-L138

---

The current code here is ad-hoc. @aliabd do you have any ideas about where we should put this code snippet better or what code snippet we can write?
(I firstly thought I could write this code at `queueing.Queue.close()`, but it seemed the `Queue` is expected to be `resume()`-ed, so I finally didn't. Was it correct?)

File: gradio/blocks.py [hunk 2]
```diff
@@ -2043,8 +2045,13 @@ def reverse(text):
...
                 requests.get(f"{self.local_url}startup-events", verify=ssl_verify)
             else:
-                pass
-                # TODO: Call the startup endpoint in the Wasm env too.
+                # NOTE: One benefit of the code above dispatching `startup_events()` via a self HTTP request is
+                # that `self._queue.start()` is called in another thread which is managed by the HTTP server, `uvicorn`
+                # so all the asyncio tasks created by the queue runs in an event loop in that thread and
+                # will be cancelled just by stopping the server.
+                # In contrast, in the Wasm env, we can't do that because `threading` is not supported and all async tasks will run in the same event loop, `pyodide.webloop.WebLoop` in the main thread.
+                # So we need to manually cancel them. See `self.close()`..
```
Comment (whitphx) [inline gradio/blocks.py:2053]: @aliabid94 Is this comment correct about the purpose of calling the HTTP endpoint `startup-events`? (Ref: https://github.com/gradio-app/gradio/pull/1969)

**Review by whitphx — COMMENTED** [review]
@aliabd The bugs should have been fixed 👍 
But I have some things to ask for better code. Appreciate your thoughts, thanks!

File: gradio/blocks.py (hunk 1 above)
Comment (aliabd) [reply gradio/blocks.py:2315]: @whitphx `Blocks.close()` only makes sense in the context of a server, because it's used to free up ports/terminate the server. 

Really, the only use case is in a notebook setting, because users can be running multiple demos sequentially in cells and could use it to reduce bloat once a demo is done. Or to run `gr.close_all()` which closes every port an Interface is running on. We do also use it in our tests. But there's no use case for it inside a demo or anything like that (we don't even document it anywhere). So it doesn't make sense in the wasm context, since there's no server or ports anyway. So I think if it's called inside wasm it shouldn't do anything (or potentially it could do something with app). You do call it in `wasm_utils.py` but I'm not sure what it's doing there? (left a comment) 

Same with `queue().close()` which is only meant to be used in the context of threads.

File: gradio/wasm_utils.py [hunk 3]
```diff
@@ -16,6 +16,8 @@ class WasmUnsupportedError(Exception):
 # the Gradio's FastAPI app instance (`app`).
 def register_app(_app):
     global app
+    if app:
+        app.blocks.close()
```
Comment (aliabd) [inline gradio/wasm_utils.py:20]: I'm not sure what this is doing since `app = None` above. Am I missing something?

File: gradio/blocks.py (hunk 1 above)
Comment (aliabd) [reply gradio/blocks.py:2315]: Let me know if this makes sense or if I'm missing something

**Comment by aliabd** [comment]
Both bugs are indeed gone! All the queue apps I'm trying are working great :) Will just do some more testing and then I'm fine with merging this. Nice work @whitphx!

**Comment by aliabd** [comment]
gradio-ui / functional-test is failing because of an issue we had that's fixed on main. You should just update the branch.

File: gradio/blocks.py (hunk 1 above)
Comment (whitphx) [reply gradio/blocks.py:2315]: Thank you for the review!

My original intension was to "clean up" all the resources and running tasks bound to the running `app` and `blocks` when a user runs new app code including `Blocks.launch()` with [`run_code()` or `run_file()`](https://github.com/gradio-app/gradio/blob/main/js/app/src/lite/index.ts#L32-L33).
The clean-up here should be something like what happens when a normal Gradio app is terminated, and I thought `Blocks.close()` was it.

Also, [`wasm_utils.register_app()`](https://github.com/gradio-app/gradio/blob/main/gradio/wasm_utils.py#L17-L19) is [called](https://github.com/gradio-app/gradio/blob/main/gradio/blocks.py#L2029) with a new FastAPI `app` object every time a user calls `Blocks.launch()`.

So I thought putting the clean-up code in `Blocks.close()` and calling it from `wasm_utils.register_app()` before registering the new `app` object would be a good solution.
If it doesn't make sense, it's ok if we can put this task cancellation code somewhere and call it on every `Blocks.launch()` call as written above. Appreciate it if you have some suggestions on it as well.

File: gradio/wasm_utils.py (hunk 3 above)
Comment (whitphx) [reply gradio/wasm_utils.py:20]: `register_app()` is called repeatedly, every time a user runs new code with `run_file()` or `run_code()`.

So,
* `app = None` is an initialization of the `app` object.
* This `if app` can be true in the second call and later.

**Comment by whitphx** [comment]
While these are not direct answers to your comment, I tried to explain what the "server" mechanism of the Wasm ver. looks like:

* As you wrote, In the Wasm ver. there is no HTTP server such as `uvicorn`. However, the FastAPI app is still working and there is also an HTTP-server-like mechanism instead.
  * In normal Gradio, the FastAPI `app` communicates with the HTTP server, `uvicorn`, via ASGI protocol (and `uvicorn` communicates with HTTP clients.)
  * In the Wasm ver., similarly, the FastAPI `app` communicates with [the JS code running in a WebWorker](https://github.com/gradio-app/gradio/blob/main/js/wasm/src/webworker/index.ts) via ASGI protocol. Pyodide automatically converts Python objects into JS objects and vice versa, so such inter-language ASGI communication is possible by writing the ASGI server in JS (its code is [here](https://github.com/gradio-app/gradio/blob/main/js/wasm/src/webworker/http.ts#L64)).
    * To do this, the FastAPI `app` object must be accessible from a global scope so that the JS code can access it via Pyodide's `pyodide.runPython()`. `wasm_utils.register_app()` and `wasm_utils.get_registered_app()` are the ad-hoc mechanism for it by [registering the `app` object from `Blocks.launch()`](https://github.com/gradio-app/gradio/blob/main/gradio/blocks.py#L2029) and [obtaining it from the JS scope](https://github.com/gradio-app/gradio/blob/main/js/wasm/src/webworker/index.ts#L134).
* In other words, in the Wasm ver., we replaced `uvicorn` with our own HTTP server implementation which communicates directly with the JS code without actual HTTP requests because the FastAPI `app` runs on Pyodide in a WebWorker and the frontend app has to communicate with it via the worker-to-main-thread messaging API, not normal HTTP.

File: gradio/blocks.py (hunk 1 above)
Comment (whitphx) [reply gradio/blocks.py:2315]: For example, putting this task cancellation code to `wasm_utils.register_app()` directly also works, instead of modifying `Blocks.close()` and calling it from `register_app()`, like
```python
def register_app(_app):
    global app

    # HERE!
    for task in asyncio.all_tasks():
        task_name = task.get_name()
        if re.match(
            r"\w+_\d+", task_name
        ):  # This regex matches the task names set by `utils.set_task_name()`
            task.cancel()

    app = _app
```

`Blocks.close()` looked like a general purpose cleanup method paired with `Blocks.launch()`, so I thought using it makes the code more robust for future changes rather than adding Wasm-specific cleanup code, but it's ok to decline it if it doesn't make sense from your aspect. Please give an advice :D

Comment (abidlabs) [reply gradio/blocks.py:2315]: > Blocks.close() looked like a general purpose cleanup method paired with Blocks.launch(), so I thought using it makes the code more robust for future changes rather than adding Wasm-specific cleanup code, but it's ok to decline it if it doesn't make sense from your aspect. Please give an advice :D

So typically when Gradio runs in a Python program, it just blocks the main thread with an empty while loop -- because if the main thread was to finish executing, it would kill the Gradio server. When a user presses CTRL+C (or the equivalent), the Gradio server stops blocking the main thread, the thread finishes, and the Gradio server ends. In this case `Blocks.close()` is NOT called. So we should not think of `Blocks.close()` as a general purpose cleanup method.

`Blocks.close()` is only used if -- for whatever reason -- the user needs to launch lots of Gradio servers in their script and then close them programmatically. This is used e.g. in our tests. In that case, calling `Blocks.close()` closes the Gradio server and frees up the port.

File: gradio/wasm_utils.py (hunk 3 above)
Comment (aliabd) [reply gradio/wasm_utils.py:20]: Got it, thanks for clarifying

File: gradio/blocks.py (hunk 1 above)
Comment (aliabd) [reply gradio/blocks.py:2315]: @whitphx Got it, I didn't realize you needed to clean up anything for `run_code` or `run_file`. In that case, It's fine to put the cleanup code in `Blocks.close()` and use it for that purpose. 

For the code itself, it's unclear to me why we need to match the task names to the ones set by `utils.set_task_name()`? Can't we just cancel all the tasks? 

Something like: 

```python
for task in asyncio.all_tasks():
    task.cancel()
```

But in any case, I'm fine with doing it your way

Comment (whitphx) [reply gradio/blocks.py:2315]: @abidlabs Thank you for the explanation. Things got very clear to me.

>  So we should not think of `Blocks.close()` as a general purpose cleanup method.

I got it.

> `Blocks.close()` is only used if -- for whatever reason -- the user needs to launch lots of Gradio servers in their script and then close them programmatically.

In turn, I think this specific usage matches the Wasm situation where we need to launch and stop different `Blocks` objects in multiple `run_file()` or `run_code()` executions keeping the same Python interpreter running.
I think `Blocks.close()` can make sense from this aspect :)

Comment (whitphx) [reply gradio/blocks.py:2315]: @aliabd Thanks!

> Can't we just cancel all the tasks?

There are some tasks which must not be cancelled, such as https://github.com/gradio-app/gradio/blob/abb5e9df47989b2c56c2c312d74944678f9f2d4e/js/wasm/src/webworker/index.ts#L109
Such tasks are mainly executed outside Gradio to manage the Gradio server across multiple `run_file()` or `run_code()` executions.
Also, I wanted to reduce risks of unexpected side effects in the future.

>  I'm fine with doing it your way

So, let me take this path, thanks!

Comment (whitphx) [reply gradio/blocks.py]: @abidlabs @aliabd Also, I created a new commit modifying the async task management to cancel the tasks without depending on the task names: https://github.com/gradio-app/gradio/pull/5124/commits/5ee60a7bc72b0f6d75a2fe9bc0e2ce312a3b8b92
which seems to be more robust.
What do you think?

Comment (aliabd) [reply gradio/blocks.py]: @whitphx I really like this, much cleaner! and no regex :) 

But this is causing an error on the console. 

```
PyodideTask exception was never retrieved
pyodide.asm.js:9 future: <PyodideTask finished name='41bqrtb8ssv_0' coro=<Queue.process_events() done, defined at /lib/python3.11/site-packages/gradio/queueing.py:425> exception=ValueError('[<gradio.queueing.Event object at 0x78dcda8>] is not in list')>
```

Comment (whitphx) [reply gradio/blocks.py]: @aliabd Thanks!

I added a fix: 9c65c010fbcbdb79d8f2a533b983a84f02cc8f5b

---

As I commented in the code, this is an existing bug.
The thrown error is not printed in the normal Python env ([Why exceptions in asnycio are late or do not appear at all? - Stack Overflow](https://stackoverflow.com/q/38875361) but it is in the Pyodide/Wasm env due to its unique event loop design.
Even in the normal env, we can confirm the same exception is raised, for example when reloading an app using `gr.load(every=...)` by adding a callback to the task created [here](https://github.com/gradio-app/gradio/blob/f09aea34d6bd18c1e2fef80c86ab2476a6d1dd83/gradio/routes.py#L594-L596) to print the exception as below.
```diff
                     blocks._queue.process_events, [event], False
                 )
                 set_task_name(task, event.session_hash, event.fn_index, batch=False)
+                def task_done_callback(task):
+                    if task.exception():
+                        print(task.exception())
+                task.add_done_callback(task_done_callback)
```
```
python run.py
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
[<gradio.queueing.Event object at 0x12dafc2e0>] is not in list
```

This implies we should fix all the code using `run_coro_in_background()` to handle exceptions (as a separate ticket of course). [The Stack Overflow answer](https://stackoverflow.com/a/38890966) says
> You should never do **fire-and-forget** technique for created tasks.

**Review by aliabd — APPROVED** [review]
Thanks @whitphx for the great work here!

I tested everything:
- queueing is working as expected in a wasm environment.
- queueing is working as expected in a non-wasm environment. (here's a [space](https://huggingface.co/spaces/aliabd/testing-fork-prs) with demos reflecting last commit)

This is good to merge! Please just update the pr from main, and run the linter so the backend tests pass (`sh scripts/format_backend.sh`)

**Comment by whitphx** [comment]
Thank you!
All tests passed.
Some deploy jobs failed but it seems because this PR is from my personal repo.

Will merge it.

**Comment by whitphx** [comment]
> (here's a [space](https://huggingface.co/spaces/aliabd/testing-fork-prs) with demos reflecting last commit)

omg, Nice space for manual tests, btw!
